### DISCUSSION TITLE
Why I think compagnies give up Freeplane : feature I miss the most

### DISCUSSION THREAD

**Discussion by Zorco1** [freeplane/freeplane#860] [General]
I'm very sorry to inform the community that a big compagny (> 70 000) is going to drop Freeplane for just one reason : the Freeplane inability to save all its map objects in one file. This compagny decision-makers argue that without this core function its impossible to share maps in the enterprise when local links or local files are involved (dead links when moving, images path lost...). I agree with them and as a Freeplane sponsor i'm hopeless.

Thats's the big issue I've raised and that I think puts a stop to the Freeplane growth. I was disappointed that the community didn't pay attention to the point.
https://github.com/freeplane/freeplane/discussions/767

All desktop mindmapping softwares in the world have the feature, except ONE : Freeplane.

I beg the most active community members to adress this critical feature.

For the compagny, I think the decision will be done in one month, may be too late.

**Comment by dpolivaev** [comment]
Unfortunately, I can't develop this feature for free because it requires too much work.
Obviously, I can not compete with nonfree solutions in some aspects.
However, if you want to become a sponsor, I could estimate costs and do it freelance-wise.

**Reply by Zorco1** [reply]
Thanks for your response : that's too bad because decision makers think Freeplane is the best but is excluded for just this reason. So I think Freeplane can compete and you can be proud of your work. I've not enough money to finance the developpement and I think the compagny has not the legal and financial tool to sponsor the developpement. But I can donate and if many do the same... I also need this feature for personal use.

**Comment by euu2021** [comment]
> This compagny decision-makers argue that without this core function its impossible to share maps in the enterprise when local links or local files are involved (dead links when moving, images path lost...)

Probably, I misunderstood your problem, but I think the relative links solve this problem, don't they? Do you know that you can use relative links in FreePlane (Preferences…->Environment->Hyperlink types->Links)?

**Reply by nnako** [reply]
Hi @euu2021 , setting relative links, alone, within the preferences would only be one necessary step not solve the problem, as the linked files would all have to be located on the same folder level or below the folder level where the mindmap itself is located. Within companies, files tend to be scattered around, which is one motivation for mindmaps to be used in the first place: build a kind of table-of-contents to be independent of actual file locations.

**Reply by euu2021** [reply]
> Within companies, files tend to be scattered around, which is one motivation for mindmaps to be used in the first place

If I understood correctly, OP wants a file format that has all the image files inside it. To me, it sounds very similar to having a folder with everything inside it (mm file + images folder). 

Now, if having relative links to other files beyond the image files is a problem, then, perhaps, an easier solution is to have relative links for the images, and absolute links for everything else.

**Comment by nnako** [comment]
Hi @Zorco1 .

My recommendation for such a big company with more than 70.000 employees would be: invest in open source development ;-) . All the big players (IBM, Microsoft, ...) do so in one or other ways. Some spend money to pay Freelance work (as @dpolivaev mentioned), some pay chosen employees to work on features valuable for the company, some find other creative ways for investment and participation.

Anyway, I would offer to build a rudimentary tool to create these MMX files, externally. Meaning that from here, further iterations could be made to finally see the "save map as MMX" within the Freeplane editor itself. These are the proposed stages:

1. implement the feature as a Python script external to FP (development version)
2. re-implement the tested feature into a Groovy addon for FP (productuon version)
3. integrate the feature into the FP Java code base (convenience version)

The first development stage would already be a usable solution. Maybe not a one-click-solution, but definitely something to solve your request. If this is acceptable for you, I would work on it on the coming weekend.

Until then, we should finalize the technical requirements for stage 1. Proposal:

- rudimentary graphical user interface to select the source mindmap and the target file name / location
- identification of all used file paths / links within the mindmap (web links and recursive handling of links to other mindmaps would be skipped in the first approach)
- localization of identified files within the file systems, then copy / paste of these files into a temporary folder
- copy / paste source mindmap into the temporary folder
- modificarion / adjustment of paths within the copied mindmap to link relatively to the copied files
- zip the folder contents as (possibly self-extracting) file to be sent and unpacket anywhere

What do you think?

**Reply by Zorco1** [reply]
Thanks a lot @nnako 
The compagny I mentionned is in the finance market and its core IT developpers are working on very specific systems very far from a mindmapping tool and use off the shelf tools.

Your offer is very interesting and would cover a need many users have (professional and personnal). How can I share my mindmaps by mail or chat without a very tricky procedure for me and the receivers. I 'd like to remind that in the mind mapping concept everything should be visual with colors (why colors have disappeared from the 1.1* Freeplane branch menu ;-)) an images. A minima one image for each branch.

https://krfofm.com/is-color-disappearing-from-our-world/(url)

https://www.waynebarry.com/2015/03/03/week-7-mind-mapping-introducing-the-concept/(url)

I use many images in my maps.

The iterations you propose are very interesting, and I think you and the community have time to developp the feature without rushing. I can help by testing developpements and send samples with images (zip files).

**Comment by nnako** [comment]
Hi folks,
  hi @Zorco1 ,

please see the new Python repository "[freeplane-packer](https://github.com/nnako/freeplane-packer)" for a solution which generates a MMX file. the final structure of an MMX file where all the stakeholders and features have been consulted, is not set, yet. this would be something to do when finalizing the functionality and carry it into the Freeplane editor.

Please tell me, what you think.

Thanks.,

**Reply by macmarrum** [reply]
@nnako thanks. Looks good. After my first test I can see is that images are missing in the mmx.zip file: 
- https://github.com/nnako/freeplane-packer/issues/1

**Reply by nnako** [reply]
solved.

please try again after updating [freeplane-python-io](https://github.com/nnako/freeplane-python-io) package and the [freeplane-packer](https://github.com/nnako/freeplane-packer) application.

Thanks.
